python/lance/types/image.py

```diff
@@ -218,3 +263,70 @@ class ImageUriScalar(pa.ExtensionScalar):
 
     def as_py(self) -> Image:
         return ImageUri(self.value.as_py())
+
+
+class ImageArray(pa.ExtensionArray):
+    @staticmethod
+    def from_pandas(obj, mask=None, type=None, safe=True, memory_pool=None):
+        if isinstance(obj, pa.ChunkedArray):
+            chunks = [ImageArray.from_pandas(c) for c in obj.chunks]
+            return pa.chunked_array(chunks, chunks[0].type)
+
+        if isinstance(obj, pa.Array):
+            if pa.types.is_binary(obj.type):
+                return ImageArray.from_storage(ImageBinary.DTYPE, obj)
+            elif pa.types.is_string(obj.type):
+                return ImageArray.from_storage(ImageUri.DTYPE, obj)
+
+        if isinstance(obj, (list, tuple, np.ndarray)) and len(obj) > 0:
+            first = obj[0]
+            if isinstance(first, Image):
+                return ImageArray.from_images(obj, mask, type, safe, memory_pool)
+            elif isinstance(first, bytes):
+                storage = pa.array(obj, mask=mask, type=pa.binary(),
+                                   safe=safe, memory_pool=memory_pool)
+                return ImageArray.from_pandas(storage)
+            elif isinstance(first, str):
+                storage = pa.array(obj, mask=mask, type=pa.string(),
+                                   safe=safe, memory_pool=memory_pool)
+                return ImageArray.from_pandas(storage)
+
+        return pa.ExtensionArray.from_pandas(
+            obj, mask=mask, type=type, safe=safe, memory_pool=memory_pool
+        )
+
+    @staticmethod
+    def from_images(images, type=None, mask=None, safe=True, memory_pool=None):
+        """
+        Create an ImageUriArray from Image instances
+        Parameters
+        ----------
+        images : sequence of Image
+        type : DataType, default None
+            If not specified then use the arrow_dtype of the first Image
+            instance
+        mask : array[bool], optional
+            Indicate which values are null (True) or not null (False).
+        safe : bool, default True
+            Check for overflows or other unsafe conversions.
+        memory_pool : pyarrow.MemoryPool, optional
+            If not passed, will allocate memory from the currently-set default
+            memory pool.
+        """
+        if len(images) > 0:
+            type = images[0].arrow_dtype
+            storage = pa.array(
+                [im.to_arrow_storage() for im in images],
+                mask=mask,
+                safe=safe,
+                memory_pool=memory_pool,
+            )
+        else:
+            storage = pa.array(images, type=type)
```

Comment:
If `images` is a sequence, this else branch means the images list is empty?

Btw, it seems the type parameter is only used here.

Reply:
i'll be more explicit and change this to pa.array([], type=type).

If the input is non-empty, then the Image instance should determine the type and not the `type` argument. wdyt?